(PIVOT ADDED) Add subsystems plus minor code changes.

BenBerol opened the pull request:
Here’s the current framework for the shooter and intake subsystems. Excluded from this pr is code that pivots the arm up and down I just havent gotten to it yet. I was thinking that ought to be in a seperate subsystem than the shooter do people agree?

BenBerol committed:
Add intake subsystem

BenBerol committed:
Add shooter subsystem and update configure objects

BenBerol committed:
Add some comments

BenBerol committed:
One more comment

ArchdukeTim commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -144,15 +179,36 @@ private void configureBindings() {
 
-    resetFieldCentricButton.onTrue(new InstantCommand((
-    ) -> {
+    resetFieldCentricButton.onTrue(new InstantCommand(() -> {
```
For commands like this that doesn't actually set motor values, you should not require them, since they don't need to interrupt whatever command is running

BenBerol replied:
Is this better?

ArchdukeTim commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -144,15 +179,36 @@ private void configureBindings() {
+    }, intakeSubsystem));
+
+    feedInButton.whileTrue(new RunCommand((
+    ) -> {
```
Some wacky formatting going on here

ArchdukeTim commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -144,15 +179,36 @@ private void configureBindings() {
+    }, shooter));
+    feedInButton.onFalse(new InstantCommand((
+    ) -> {
+      shooter.feed(0);
```
I recommend making a default command of the feeder subsystem that sets the motor speed to 0

ArchdukeTim commented on src/main/java/frc/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -0,0 +1,37 @@
+    intakeMotor.set(speed);
+  }
+
+  public Command exampleMethodCommand() {
+    // Inline construction of command goes here.
+    // Subsystem::RunOnce implicitly requires `this` subsystem.
+    return runOnce(
+        () -> {
+          /* one-time action goes here */
+        });
+  }
```
Don't leave the template code in

BenBerol commented:
Ok thanks for the feedback, and @ArchdukeTim do you agree that an arm pivot should have it's own subsystem?

ArchdukeTim commented:
Yes. Anything that can run at the same time as another subsystem should be its own subsystem

BenBerol committed:
Tim's changes

BenBerol commented:
Ok so I turned the resetFieldCentric into a a swerveDrive command without requirements, which I think is what you wanted, added a default command to the shooter and intake to zero them, and removed some template code!

ArchdukeTim commented on src/main/java/frc/robot/subsystems/SwerveDriveSubsystem.java:
```diff
@@ -138,9 +138,11 @@ public Command toggleFieldCentric() {
-        odometry.zeroHeading();
-        resetLockRot();
+    public Command resetFieldCentric() {
+        return runOnce(() -> {
```
`runOnce` actually requires the subsystem by default, since this is the `SubsystemBase` version. To not require it, use `Commands.runOnce`

ArchdukeTim commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -177,11 +167,17 @@ private void configureBindings() {
       
     }, swerveDrive));
 
+    shooter.setDefaultCommand(new RunCommand(() -> {
+      shooter.feed(0);
+    }, shooter));
+
+    intakeSubsystem.setDefaultCommand(new RunCommand(() -> {
+      intakeSubsystem.intake(0);
+    }, intakeSubsystem));
+
```
This is fine. Alternatively, in the constructor for these subsystems you can do

```
setDefaultCommand(run(() -> feed(0)));

setDefaultCommand(run(() -> intake(0)));
```

ArchdukeTim replied:
Likewise, all of the `new RunCommands` _could be_ converted to use the command factory methods. Not necessary, just a little simpler code

BenBerol replied:
I like leaving these default commands in the robot container class so that they are next to all of the inputted commands, I think it makes the most sense even though that code is definitely cleaner.

BenBerol committed:
Remove command requirements

BenBerol commented:
Ok I've updated the runOnce commands

BenBerol committed:
Add pivot subsystem

BenBerol committed:
Add nt entry

BenBerol committed:
More pivot work and creating feed and climber subsystems

ArchdukeTim commented on src/main/java/frc/robot/subsystems/PivotSubsystem.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    public void setPivotPosition(double pos) {
+        pivotMotor.setVoltage(armFeedforward.calculate(pos, 0) + pivotPidController.calculate(ntPivotPosition.getDouble(0), pos));
```
Don't read your current position from network tables!

BenBerol replied:
On it!

ArchdukeTim commented on src/main/java/frc/robot/subsystems/PivotSubsystem.java:
```diff
@@ -0,0 +1,83 @@
+    private ArmFeedforward armFeedforward = new ArmFeedforward(0, 0, 0);
+
+    private double targetPos;
+    private double lockPos;
```
What is `lockPos`?

BenBerol replied:
Lock pos is the position that we hold when we are changing target pos. It lets us adjust target pos and see the value we're moving to while holding a different position

BenBerol replied:
I don't think we'll have lockPos forever but while testing it streamlines the process and allows us to be more controlled

ArchdukeTim reviewed (changes requested):
Rather than repeatedly calling the setPosition method with the `lockPos` as your default command, your commands should change `targetPos`, and then the default command should just set the arm position to `targetPos`. No argument needed for that method. 

I also recommend using the on-board PID controller for the spark max, and you can set an arbitrary feedforward voltage with `getPIDController().setReference(targetPos, kPosition, 0, arbFf)`

BrandonWerbel commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -113,21 +137,37 @@ public void configureObjects() {
-    JoystickButton turtleButton = new JoystickButton(rightJoystick, 1);
+    JoystickButton turtleButton = new JoystickButton(rightJoystick, 3);
 
     JoystickButton fieldCentricButton = new JoystickButton(rightJoystick, 2);
```
Small thing, but I would think about grouping these button declarations by subsystem or something, instead of giving each it's own code block, simply for better readability.

BenBerol replied:
Yeah I was doing something similar but it wasn't full comprehensive, I'll finish it

BrandonWerbel commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -78,26 +106,22 @@ public class RobotContainer {
 
     SlewRateLimiter limitX = new SlewRateLimiter(6);
     SlewRateLimiter limitY = new SlewRateLimiter(6);
+    SlewRateLimiter limitI = new SlewRateLimiter(6);
```
Can you put a comment on what these SlewRateLimiter's are for? `limitI` and `limitS` are not very informative.

BrandonWerbel replied:
Also is there a reason you need them in the first place? Generally SlewRateLimiter is to limit acceleration, but I would think you want shooter and intake to speed up as fast as possible

BenBerol replied:
Sure I'll add comments. We originally didn't have limiters for the intake and shooter but then running the shooter from 0 to 50+ percent would kill comms

ArchdukeTim replied:
You should be able to set the rampRate on the motor controller directly.

BenBerol commented:
> I also recommend using the on-board PID controller for the spark max, and you can set an arbitrary feedforward voltage with `getPIDController().setReference(targetPos, kPosition, 0, arbFf)`

Is this necessary? I don't see a setReference that takes in a position and it is also unintuitive for this code because we realized we have to invert the PID values but not the feedforward.

BenBerol committed:
Some requested changes and more nt entries

ArchdukeTim commented on src/main/java/frc/robot/subsystems/PivotSubsystem.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    public void setPivotPosition(double pos) {
+        pivotMotor.setVoltage(armFeedforward.calculate(pos, 0) - pivotPidController.calculate(pivotEncoder.getPosition(), pos));
```
Why are you subtracting the feedback? 

Also, feedforward should use the current position, rather than the target position, since gravity acts on where you are now, rather than where you want to be

BenBerol replied:
We are subtracting the feedback because our encoder is on so that a positive voltage decreases the encoder value when I believe the PID is set so that a positive voltage encreases encoder values

BenBerol replied:
I'm pretty sure that feedforward is supposed to use the target location of the arm for setting. In this [tuning simulator](https://docs.wpilib.org/en/stable/docs/software/advanced-controls/introduction/tuning-vertical-arm.html) the feedforward is controlled by the setpoint of the arm and in the [Feedforward docs ](https://docs.wpilib.org/en/stable/docs/software/advanced-controls/introduction/introduction-to-feedforward.html#introduction-to-dc-motor-feedforward) it says that feedforward should take in values from the setpoint. From what I've read I believe feedforward is a voltage that is set once per setpoint and isn't supposed to change until the set point does which would be undermined by taking in a changing current position. I could be wrong but I think the idea is that the arm should naturally not have the correct voltage in it's current position to stay steady causing it to slowly fall or raise to the set point as long as it's in the right quadrant of the circle

ArchdukeTim replied:
> encoder is on so that a positive voltage decreases the encoder value 

You should invert your encoder. You can do this on the spark max.

> feedforward is supposed to use the target location of the arm for setting.

Yeah if you only call `setPivotPosition` once, you're correct it should be on the goal position. Ideally you would continuously update the feedforward to use where the arm currently is.

BenBerol replied:
Are you sure? In the tuning simulator the FF doesn't use the position of the arm and I don't think that feedforward voltage is supposed to vary until the setpoint does since it's a constant voltage that in a perfect world would move the arm to its setpoint.

BenBerol replied:
In this [article](https://www.ctrlaltftc.com/feedforward-control) discussing feedforward they give an example code of feedforward:
double Kcos = SOME_VALUE_YOU_TUNE;
double reference = YOUR_TARGET_ANGLE;
double power = Math.cos(reference) * Kcos;

ArchdukeTim replied:
As the arm moves, the effect of gravity changes, and thus the feedforward to overcome that gravity also needs to change. It's not a huge issue since your PID control will eventually overcome that. But even in a perfect world the force gravity would still change magnitude as the angle between the arm and the ground changes

BenBerol replied:
Yeah I think that's the idea of the PID controls is that they solve that problem with feedforward, so it's part of the design that they correct for the imperfect FF voltage. All of the explanations I see online are setting one FF voltage relative to the target angle of the arm. Any difference between the force required at a different angle and the force provided by FF should be fixed by the feedback voltage

ArchdukeTim commented:
> Is this necessary

No, but you get a free 1kHz control loop instead of 50 Hz

> I don't see a setReference that takes in a position

https://codedocs.revrobotics.com/java/com/revrobotics/sparkpidcontroller#setReference(double,com.revrobotics.CANSparkBase.ControlType,int,double)

As long as you use `ControlType.kPosition` you'll use positional PID

> we have to invert the PID values

What do you mean invert?

BenBerol commented:
> No, but you get a free 1kHz control loop instead of 50 Hz

Oh that does sound nice, I'll try and convert to this then

https://codedocs.revrobotics.com/java/com/revrobotics/sparkpidcontroller#setReference(double,com.revrobotics.CANSparkBase.ControlType,int,double)
> 
> As long as you use `ControlType.kPosition` you'll use positional PID

Ok, so it will naturally take in a current position value?

> > we have to invert the PID values
> 
> What do you mean invert?

I just mean that like I commented above we have to subtract feedback to account for a backward encoder. However, the Sparkmax pid method uses the integrated encoder, right? So this ought to not be an issue at all

Also, one other question is the setReference that takes in 4 parameters that you linked better to use than the value and control type only version of the method? I don't think we need an arbFF since we will have a FF built into the controller and the pidSlot can be automatically set to 0 so it seems like these parameters are unnecessary

BenBerol commented:
Ok in summary, I'll try using the built in SparkMax pid controls instead of the manual PID controls, but if I have issues with it I think I will stick with the current system to save time since I doubt an increase in Hz will noticeably improve performance

ArchdukeTim commented:
> we will have a FF built into the controller and the pidSlot can be automatically set to 0 so it seems like these parameters are unnecessary

The ff built into the spark max doesn't work well for position because feedforward is strictly correlates velocity/acceleration (and offsets gravity), so I would not recommend using it for the arm

BenBerol commented:
> > we will have a FF built into the controller and the pidSlot can be automatically set to 0 so it seems like these parameters are unnecessary
> 
> The ff built into the spark max doesn't work well for position because feedforward is strictly correlates velocity/acceleration (and offsets gravity), so I would not recommend using it for the arm

Our only use of feedforward right now is to offset gravity so wouldn't this accomplish that?

ArchdukeTim commented:
No because the spark max just does kFF * set point (linear), but the equation to offset gravity is kG * cos(angle) (sinusoidal).

BenBerol commented:
> No because the spark max just does kFF * set point (linear), but the equation to offset gravity is kG * cos(angle) (sinusoidal).

Right ok, so then would you recommend setting an arbFF equal to kG * cos(angle)

BenBerol commented:
Oh one issue I just thought of was that the encoder built into the motor won't correlate well to the rotation of our arm because the chain we use isn't perfect and there is some noise in the system so that the rotation of the motor won't correlate correctly to the rotation of the arm so unless there's a setting to take in the absolute encoder value then we won't be able to use the built in pid controller

BenBerol committed:
Update feed speed

BenBerol commented:
Ok I'm going to merge unless there are any other crucial changes
